File: orig/mutated_maze-0.5/gamelib/fog.py

import sys
import os
import random
import pygame
import pygame.image
import itertools

import data
from gamelib import ega


FOGFILES=(
    ('fog01.png', 'fog01_inverse.png'),
    ('fog02.png', 'fog02_inverse.png'),
    ('fog03.png', 'fog03_inverse.png'),
    ('fog04.png', 'fog04_inverse.png'),
    ('fog05.png', 'fog05_inverse.png'),
    ('fog06.png', 'fog06_inverse.png'),
)

# setup a random list of colours
maze_colours = list(ega.SPRITE_COLOURS)
random.shuffle(maze_colours)
colour_iterator = itertools.cycle(maze_colours)

class Fog(object):

    def __init__(self, gamestate):
        '''
        :type gamestate: GameState
        '''

        self.gamestate = gamestate
        self.for_maze = gamestate.maze

        (fog_name, inverse_name) = random.choice(FOGFILES)
        self.fog_surface = self.load_fog(fog_name)
        self.inverse_surface = self.load_fog(inverse_name)

        self.surface = self.fog_surface

        self.mutate_snd = pygame.mixer.Sound(data.filepath("mutate.wav"))

        self.x = gamestate.maze.width_px
        self.y = 0
        self.vx = -1
        self.vy = 0

        self.wanted_direction=None
        self.direction=None

        self.passed = False

        # figure this out before hand.
        #self.mutate_point = random.randint(-self.for_maze.width_px/2, self.for_maze.width_px/2)

        self.mutated = False

        self.mutate_count = random.randint(self.for_maze.width_px*0.5, self.for_maze.width_px*1.5)
        self.flash_count = 0



    def load_fog(self, path):
        maze = self.gamestate.maze
        datapath = data.filepath(path)
        i = pygame.image.load(datapath)
        i = pygame.transform.scale(i, (maze.width_px, maze.height_px))
        i = i.convert_alpha()
        return i

    def mutate(self):
        self.mutate_snd.play()

        maze = self.gamestate.maze

        selected_nodes = []

        new_maze_colour = colour_iterator.next()

        # select nodes under the fog
        for node in maze._data:
            c_x = node.x*16 + 8 # mid point
            c_y = node.y*16 + 8
            if self.pos_covered_with_fog((c_x, c_y)):
                node.colour = new_maze_colour
                selected_nodes.append(node)

        # find all the mobs and move them, also mutate them
        mutate_mobs = []
        for mob in self.gamestate.alive_mobs:
            c_x = mob.x+8
            c_y = mob.y+8
            if self.pos_covered_with_fog((c_x, c_y)):
                mutate_mobs.append(mob)

        random.shuffle(selected_nodes)

        for mob,node in zip(mutate_mobs, itertools.cycle(selected_nodes)):
            mob.do_fog_mutate(node.pos_px)

        self.for_maze.regenerate_selected(selected_nodes)




    def pos_covered_with_fog(self, pos):
        fog_x = pos[0]-self.x
        fog_y = pos[1]-self.y

        if fog_x < 0 or fog_x >= self.gamestate.maze.width_px:
            return False
        if fog_y < 0 or fog_y >= self.gamestate.maze.height_px:
            return False

        c = tuple(self.fog_surface.get_at((fog_x, fog_y)))
        return c[3] > 0

    def update(self):
        self.x += self.vx
        if self.x < -self.for_maze.width_px:
            self.passed = True

        if not self.mutated:
            self.mutate_count -= 1

            if self.mutate_count < 0:
                self.mutate()
                self.mutated = True
                self.surface = self.inverse_surface

            # don't flash too early.
            elif self.mutate_count < 200:
                self.flash_count += 10
                if self.flash_count > self.mutate_count:
                    self.flash_count = 0
                    if self.surface is self.inverse_surface:
                        self.surface = self.fog_surface
                    else:
                        self.surface = self.inverse_surface





    def render(self, to_surface):
        to_surface.blit(self.surface, (self.x,self.y))

File: fog.go
package main

import (
	"math/rand"

	"github.com/qeedquan/go-media/sdl"
	"github.com/qeedquan/go-media/sdl/sdlmixer"
)

var fogFiles = [][2]string{
	{"fog01.png", "fog01_inverse.png"},
	{"fog02.png", "fog02_inverse.png"},
	{"fog03.png", "fog03_inverse.png"},
	{"fog04.png", "fog04_inverse.png"},
	{"fog05.png", "fog05_inverse.png"},
	{"fog06.png", "fog06_inverse.png"},
}

var mazeColors []sdl.Color

func init() {
	mazeColors = append(mazeColors, SpriteColors...)
}

type Fog struct {
	Blitter
	gs          *GameState
	mutateSound *sdlmixer.Chunk

	mutateCount int
	flashCount  int

	pal     int
	vx, vy  int
	passed  bool
	mutated bool
	fog     *Image
	inverse *Image
}

func newFog(gs *GameState) *Fog {
	f := &Fog{}
	f.gs = gs

	for i := len(mazeColors) - 1; i >= 1; i-- {
		j := rand.Intn(i + 1)
		mazeColors[i], mazeColors[j] = mazeColors[j], mazeColors[i]
	}

	fogChoice := fogFiles[rand.Intn(len(fogFiles))]
	f.fog = loadFog(fogChoice[0], gs.maze)
	f.inverse = loadFog(fogChoice[1], gs.maze)
	f.image = f.fog

	f.mutateSound = loadSound("mutate.wav")
	f.x = gs.maze.Px()
	f.vx = -1
	f.mutateCount = f.x/2 + rand.Intn(f.x*4/3-f.x/2)
	return f
}

func loadFog(name string, m *Maze) *Image {
	return loadImage(name, scaleImage(m.Px(), m.Py()))
}

func (f *Fog) mutate() {
	playSound(f.mutateSound)

	newMazeColor := mazeColors[f.pal]
	f.pal = (f.pal + 1) % len(mazeColors)

	var selectedNodes []*MazeNode
	for _, n := range f.gs.maze.nodes {
		cx := n.x*16 + 8
		cy := n.y*16 + 8
		if f.posCoveredWithFog(cx, cy) {
			n.color = newMazeColor
			selectedNodes = append(selectedNodes, n)
		}
	}

	var mutateMobs []*Mob
	for _, m := range f.gs.AliveMobs() {
		cx := m.x + 8
		cy := m.y + 8
		if f.posCoveredWithFog(cx, cy) {
			mutateMobs = append(mutateMobs, m)
		}
	}

	for i := len(selectedNodes) - 1; i >= 1; i-- {
		j := rand.Intn(i + 1)
		selectedNodes[i], selectedNodes[j] = selectedNodes[j], selectedNodes[i]
	}

	i := 0
	for _, m := range mutateMobs {
		n := selectedNodes[i]
		i = (i + 1) % len(selectedNodes)
		x, y := n.Pxy()
		m.FogMutate(x, y)
	}

	f.gs.maze.RegenSelected(selectedNodes)
}

func (f *Fog) posCoveredWithFog(x, y int) bool {
	fx := x - f.x
	fy := y - f.y

	if fx < 0 || fx >= f.gs.maze.Px() {
		return false
	}
	if fy < 0 || fy >= f.gs.maze.Py() {
		return false
	}

	return f.fog.alpha.AlphaAt(fx, fy).A > 0
}

func (f *Fog) Update() {
	f.x += f.vx
	if f.x < -f.gs.maze.Px() {
		f.passed = true
	}

	if !f.mutated {
		if f.mutateCount--; f.mutateCount < 0 {
			f.mutate()
			f.mutated = true
			f.image = f.inverse
		} else if f.mutateCount < 200 {
			f.flashCount += 10
			if f.flashCount > f.mutateCount {
				f.flashCount = 0
				if f.image == f.inverse {
					f.image = f.fog
				} else {
					f.image = f.inverse
				}
			}
		}
	}
}

func (f *Fog) Free() {
	f.fog.Free()
	f.inverse.Free()
}
